File: global/front_desk/mis_issue_tracking/mail_for_changes.py
#!/usr/bin/python

#----------------------------------------------------------
#
#	Watches the notification table for the
#	problem resolution database, and mails
#	notifications to those that need them.
#
#----------------------------------------------------------

import MySQLdb
import os
import string
import sys
import re
import time

sqlhost="reddwarf"
sqluser="jeffb"
sqlpass="bongsrolinge"
sqldata="global"

tomorrow=time.strftime("%Y-%m-%d",time.localtime(time.time() + 86400))
yesterday=time.strftime("%Y-%m-%d",time.localtime(time.time() - 86400))
day_of_week=time.strftime("%A",time.localtime(time.time()))

#
#	E-mail Message Stuff
#
mailprettyname="IT Issue Notifier"
mailreturn="webmaster@umci.com"

if (len(sys.argv)>1):
	issue_id = sys.argv[1]
	issue_id_string=issue_id
	#issue_id_string='%(issue_id)d' % vars()
else:	
	print "ERROR: No issue id specified!"
	sys.exit()
        

# This had better be sendmail, or something *just* like it..
# (and you better be a trusted user)
sendmail="/usr/sbin/sendmail"

it_issuedb=MySQLdb.connect(host=sqlhost,user=sqluser,passwd=sqlpass,db=sqldata)

def mail_them(contact,issue):
	issue_id=issue['issue_id']
	issue_id_string='%(issue_id)d' % vars()
	mimetag='''------------=_3E260B33.3708D0F9\r\n'''
	mailcommand=sendmail + " -f " + mailreturn + " -F \"UMC IT Issue Notifier\" -- " + contact['email']
	mailpipe=os.popen(mailcommand,'w')

	mailpipe.write("From: UMC IT Issue Notifier <webmaster@umci.com>\r\nTo: " + contact['first_name'] + " " + contact['last_name'] + " <" + contact['email'] + ">\r\n" + "Subject: IT Issue Modified (" + issue_id_string + ")\r\n")
	mailpipe.write('''Content-Type: text/html; charset=utf-8\r\n''')
	mailpipe.write('''Content-Description: Automatic notification from UMC Mail System\r\n''')
	mailpipe.write('''Content-Disposition: inline\r\n''')
	mailpipe.write('''Content-Transfer-Encoding: 8bit\r\n''')
	
	requested_completion=issue['requested_completion']
	rcm=requested_completion.month
	rcd=requested_completion.day
	rcy=requested_completion.year
	rcms='%(rcm)d' % vars()
	rcds='%(rcd)d' % vars()
	rcys='%(rcy)d' % vars()
	requested_completion_string=rcms + "-" + rcds + "-" + rcys


	expected_completion=issue['expected_completion']
	ecm=expected_completion.month
	ecd=expected_completion.day
	ecy=expected_completion.year
	ecms='%(ecm)d' % vars()
	ecds='%(ecd)d' % vars()
	ecys='%(ecy)d' % vars()
	expected_completion_string=ecms + "-" + ecds + "-" + ecys

	action_completion=issue['expected_completion_date']
	acm=action_completion.month
	acd=action_completion.day
	acy=action_completion.year
	acms='%(acm)d' % vars()
	acds='%(acd)d' % vars()
	acys='%(acy)d' % vars()
	action_completion_string=acms + "-" + acds + "-" + acys
	
	if issue['us_or_them']=="them":
		issue['us_or_them']="<font color=red><b>USERS (that's you)</b></font>"
	else:
		issue['us_or_them']="<font color=green>IT</font>"
		
	message="""
	<h2>IT Issue """ + issue_id_string + """ Updated</h2><hr>
	<a href=http://pipeline.umci.com/global/mis_issue_request/?mode=issue_edit&issue_id=""" + issue_id_string + """><font color=blue>Load this issue</font></a><br>
	<b>New issue information:</b>
	<table border=1>
	<tr><td>
		Subject
	</td><td>
		""" + issue['name'] + """
	</td></tr>

	<tr><td>
		Original Description
	</td><td>
		""" + issue['description'] + """ 
	</td></tr>

	<tr><td>
		Requested Completion Date
	</td><td>
		""" + requested_completion_string + """
	</td></tr>

	<tr><td>
		Expected Completion Date
	</td><td>
		""" + expected_completion_string + """
	</td></tr>

	<tr><td>
		Next Action Date
	</td><td>
		""" + action_completion_string + """
	</td></tr>

	<tr><td>
		Status
	</td><td>
		""" + issue['status'] + """
	</td></tr>

	<tr><td>
		Notes
	</td><td>
		""" + issue['notes'] + """
	</td></tr>

	<tr><td>
		Next Action
	</td><td>
		""" + issue['action_required'] + """
	</td></tr>

	<tr><td>
		Next Action Responsibility
	</td><td>
		""" + issue['us_or_them'] + """
	</td></tr>

	</table>
	"""


	mailpipe.write(message + "\r\n")
	mailpipe.close()


def mail_us(contact,issue):
	issue_id=issue['issue_id']
	issue_id_string='%(issue_id)d' % vars()
	mimetag='''------------=_3E260B33.3708D0F9\r\n'''
	mailcommand=sendmail + " -f " + mailreturn + " -F \"UMC IT Issue Notifier\" -- " + contact['email']
	mailpipe=os.popen(mailcommand,'w')

	mailpipe.write("From: UMC IT Issue Notifier <webmaster@umci.com>\r\nTo: " + contact['first_name'] + " " + contact['last_name'] + " <" + contact['email'] + ">\r\n" + "Subject: IT Issue Modified (" + issue_id_string + ")\r\n")
	mailpipe.write('''Content-Type: text/html; charset=utf-8\r\n''')
	mailpipe.write('''Content-Description: Automatic notification from UMC Mail System\r\n''')
	mailpipe.write('''Content-Disposition: inline\r\n''')
	mailpipe.write('''Content-Transfer-Encoding: 8bit\r\n''')
	
	requested_completion=issue['requested_completion']
	rcm=requested_completion.month
	rcd=requested_completion.day
	rcy=requested_completion.year
	rcms='%(rcm)d' % vars()
	rcds='%(rcd)d' % vars()
	rcys='%(rcy)d' % vars()
	requested_completion_string=rcms + "-" + rcds + "-" + rcys


	expected_completion=issue['expected_completion']
	ecm=expected_completion.month
	ecd=expected_completion.day
	ecy=expected_completion.year
	ecms='%(ecm)d' % vars()
	ecds='%(ecd)d' % vars()
	ecys='%(ecy)d' % vars()
	expected_completion_string=ecms + "-" + ecds + "-" + ecys


	action_completion=issue['expected_completion_date']
	acm=action_completion.month
	acd=action_completion.day
	acy=action_completion.year
	acms='%(acm)d' % vars()
	acds='%(acd)d' % vars()
	acys='%(acy)d' % vars()
	action_completion_string=acms + "-" + acds + "-" + acys

	if issue['us_or_them']=="them":
		issue['us_or_them']="<font color=green>USERS</font>"
	else:
		issue['us_or_them']="<font color=red><b>IT</b></font>"
		
	message="""
	<h2>IT Issue """ + issue_id_string + """ Updated</h2><hr>
	<a href=http://pipeline.umci.com/global/mis_issue_tracking/?mode=issue_edit&issue_id=""" + issue_id_string + """><font color=blue>Load this issue</font></a><br>
	<b>New issue information:</b>
	<table border=1>
	<tr><td>
		Subject
	</td><td>
		""" + issue['name'] + """
	</td></tr>

	<tr><td>
		Original Description
	</td><td>
		""" + issue['description'] + """ 
	</td></tr>

	<tr><td>
		Requested Completion Date
	</td><td>
		""" + requested_completion_string + """
	</td></tr>

	<tr><td>
		Expected Completion Date
	</td><td>
		""" + expected_completion_string + """
	</td></tr>

	<tr><td>
		Next Action Date
	</td><td>
		""" + action_completion_string + """
	</td></tr>

	<tr><td>
		Status
	</td><td>
		""" + issue['status'] + """
	</td></tr>

	<tr><td>
		Notes
	</td><td>
		""" + issue['notes'] + """
	</td></tr>

	<tr><td>
		Next Action
	</td><td>
		""" + issue['action_required'] + """
	</td></tr>

	<tr><td>
		IT Notes
	</td><td>
		""" + issue['it_notes'] + """
	</td></tr>

	<tr><td>
		Next Action Responsibility
	</td><td>
		""" + issue['us_or_them'] + """
	</td></tr>

	</table>
	"""


	mailpipe.write(message + "\r\n")
	mailpipe.close()


creator_notified=0

query=""" select * from mis_issue_index,mis_issue_history where mis_issue_index.issue_id = '""" + issue_id_string + """' and mis_issue_index.issue_id = mis_issue_history.issue_id and current_hist_item = 'Y' and email_sent = 'N'"""
issues=it_issuedb.cursor(MySQLdb.cursors.DictCursor)
issues.execute(query)

if (issues.rowcount > 1):
	print "ERROR: more than 1 current item for that issue.. that's not good!"
	sys.exit()

if (issues.rowcount < 1):
	print "ERROR: no current item, or mail already sent!"
	sys.exit()

issue_list=issues.fetchall()
issue=issue_list[0]


issue_history_id=issue['issue_history_id']
issue_history_id_string='%(issue_history_id)d' %vars()

query="select * from mis_issue_concerned_parties,contacts where issue_history_id = '" + issue_history_id_string + "' and mis_issue_concerned_parties.contacts_id = contacts.contacts_id"
issues.execute(query)
userlist=issues.fetchall()


for user in userlist:
	if user['contacts.contacts_id']==issue['creator']:
		creator_notified=1
	if user['contacts.contacts_id']!=issue['entered_by']:
		if user['us_or_them']=='us':
			mail_us(user,issue)
		else:
			mail_them(user,issue)

if creator_notified==0:
	print "creator not seen.. might want to look him/her up and notify him/her"

query="update mis_issue_history set email_sent = 'Y' where issue_history_id = '" + issue_history_id_string + "'"
issues.execute(query)

sys.exit()

File: global/front_desk/mis_issue_tracking/mail_for_changes.php
#!/usr/local/bin/php
<?

require("global_connect_rw.phtml");
require('querylib.inc');
$site="pipeline.umci.com";
$issue_id=addslashes($argv[1]);
$exclude_user_id=addslashes($argv[2]);
$contacts_view_url="http://$site/global/contacts/contactview.phtml?contacts_id=";

function mail_contact($contact,$issue_info) {
	global $site,$contacts_view_url;
	$index_info=getone("select * from mis_issue_index where issue_id = '$issue_info->issue_id'");
	$contact=addslashes($contact);
	$itisus=getoneb("select * from mis_issue_uslist where contacts_id = '$contact'");
	$rcd=invali_date($issue_info->requested_completion,'/');
	$ecd=invali_date($issue_info->expected_completion,'/');
	$nad=invali_date($issue_info->expected_completion_date,'/');
	$created_date=invali_date($issue_info->issue_create_date);
	
	
	$people_res=@mysql_query("select * from mis_issue_concerned_parties left join contacts using (contacts_id) where issue_history_id = '$issue_info->issue_history_id'");
	$us_people="";
	$them_people="";
	while ($people_row=@mysql_fetch_object($people_res)) {
		if ($people_row->us_or_them=="us") $us_people=$us_people . "<a target=contact_view_win href=$contacts_view_url$people_row->contacts_id><font color=blue>$people_row->display_name</font></a><br>";
		if ($people_row->us_or_them=="them") $them_people=$them_people . "<a target=contact_view_win href=$contacts_view_url$people_row->contacts_id><font color=blue>$people_row->display_name</font></a><br>";
		}
	if ($itisus) {
		$issue_url="http://pipeline.umci.com/global/mis_issue_tracking/?mode=issue_edit&issue_id=$issue_info->issue_id";
		} else {
		$issue_url="http://pipeline.umci.com/global/mis_issue_request/?mode=issue_edit&issue_id=$issue_info->issue_id";
		}
	
	if ($index_info->status=="complete") {
		$action_text="completed";
		//if (!($itisus)) $additional_message="This item will be considered closed unless you follow the link to edit the item and then reject it. We would also appreciate it if when you are completely satisfied that the issue has been resolved to your satisfaction that you log in, and approve the issues completion.<p>";
		$additional_message="This item will be considered closed unless you follow the link to edit the item and then reject it. We also ask that you log in and approve this item when you are satisfied with it's resolution.<p>";
		//else $additional_message="";
		} else {
		$action_text="updated";
		$additional_message="";
		}
	$creator_info=getoneb("select * from contacts where contacts_id = '$issue_info->creator'");
	$subject="IT issue $action_text";
	$message="<b>IT Issue number $issue_info->issue_id has been $action_text</b><p>$additional_message
	
	<a href=\"$issue_url\"><font color=blue>Edit Issue $issue_info->issue_id</font></a>
	
	
	<table border=1 cellspacing=0 cellpadding=4><tr><td bgcolor=#3355ff>
	<table border=1 cellspacing=0 cellpadding=4 bgcolor=white>
	<tr bgcolor=#99aaff><td colspan=2 align=center>
		<b>$issue_info->name (<a href=\"$issue_url\"><font color=blue>$issue_info->issue_id</font></a>)</b>
	</td></tr>
	
	<tr><td width=50%>
		<b>Description</b>
	</td><td width=50%>
		$issue_info->description
	</td></tr>

	<tr><td>
		<b>Issue Creator</b>
	</td><td>
		<a target=contact_view_win href=$contacts_view_url$creator_info->contacts_id><font color=blue>$creator_info->display_name</font></a>
	</td></tr>
	
	<tr><td>
		<b>Issue Created on
	</td><td>
		$created_date
	</td></tr>

	<tr><td>
		<b>Item Status</b>
	</td><td>
		$issue_info->status
	</td></tr>
	
	<tr><td>
		<b>Requested&nbsp;Completion&nbsp;Date</b>
	</td><td>
		$rcd
	</td></tr>
		
	<tr><td>
		<b>Expected Completion Date</b>
	</td><td>
		$ecd
	</td></tr>
		
	<tr><td valign=top>
		<b>Activity/Notes</b> <i><font size=-1>($nad)</font></i>
	</td><td>
		$issue_info->action_required
	</td></tr>

	<tr><td>
		<b>Additional Notes</b>
	</td><td>
		$issue_info->notes&nbsp;
	</td></tr>";
	
	if ($itisus) {
		$message=$message . "
		<tr><td>
			<b>IT Notes</b>
		</td><td>
			$issue_info->it_notes&nbsp;
		</td></tr>
		";
		}
	
	if ($issue_info->us_or_them=="us") $nar=$us_people;
	if ($issue_info->us_or_them=="them") $nar=$them_people;
	
	if ($index_info->status!="complete") {
	$message = $message . "
	<tr><td valign=top>
		<b>Next action responsibility</b>
	</td><td>
		$nar
	</td></tr>";
	}		
	$message = $message . "
	<tr bgcolor=#99aaff><td align=center>
		<b>Assigned IT People</b>
	</td><td align=center>
		<b>Users</b>
	</td></tr>	
	
	<tr><td valign=top>
		$us_people
	</td><td valign=top>
		$them_people
	</td></tr>
	
	</table>
	</td></tr></table>
	";
	
	//echo $message;
	queue_email($contact,$subject,$message,'Y');
	
	}

$query="select * from mis_issue_index,mis_issue_history 
	where 
	mis_issue_index.issue_id = '$issue_id' and 
	mis_issue_index.issue_id = mis_issue_history.issue_id and 
	current_hist_item = 'Y' and email_sent = 'N'";
$issue_info=getoneb($query);


if ($issue_info) {
	$query="select * from mis_issue_concerned_parties,contacts 
		where 
		issue_history_id = '$issue_info->issue_history_id' and 
		mis_issue_concerned_parties.contacts_id = contacts.contacts_id group by contacts.contacts_id";
	$res=@mysql_query($query);
	
	while ($row=@mysql_fetch_object($res)) {
		//echo "<hr>$row->display_name";
		if ($exclude_user_id!=$row->contacts_id) mail_contact($row->contacts_id,$issue_info);
		}
	@mysql_query("update mis_issue_history set email_sent = 'Y' where issue_history_id = '$issue_info->issue_history_id'");
	
	}
?>
